### PR TITLE
Add GitHub action to format and lint code

### PR DESCRIPTION
**Pull request by dyastremsky**
This pull request adds pre-commit hooks. It also includes running the pre-commit hooks on these files, plus making fixes based on them (e.g. making sure all `.py` and `.sh` executables are set to be executable and have shebangs, spelling corrections, etc.).

This pull request adds pre-commit  hooks to run the following:
- Black for formatting Python
- Flake8 for linting Python for conformance to PEP8, PyFlakes, and circular complexity (max-line-length set to match Black's 88 default/recommended character limit; this and other settings from Black's documentation [here](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#line-length) to avoid enforcement of limit when unnecessary).
- isort for deterministically sorting and organizing Python imports
- Clang for formatting C++, C, Java, TextProto, Proto, and CUDA
- Codespell for finding common spelling mistakes 

In addition, this adds these native pre-commit hooks, details [here](https://pre-commit.com/hooks.html):
- check-case-conflict
- check-executables-have-shebangs
- check-merge-conflict
- check-json
- check-toml
- check-yaml
- check-shebang-scripts-are-executable
- end-of-file-fixer
- mixed-line-ending
- requirements-txt-fixer
- trailing-whitespace

To run these locally, you can go into the repo and run `pip install pre-commit` and `pre-commit install`. After that has been done once, you just need to run `pre-commit run --all-files` any time you want to apply them. Once installed, it should run automatically on commit for files included in a commit. To correct spelling errors found by codespell when possible, you can install codespell (`pip install codespell` and potentially `pip install --upgrade codespell` to use the toml file in the directory) and call `codespell -w <path>`.

This pull request also adds a GitHub action so that these are run for every pull request and push to main. The core changes are in pre-commit.yml, pre-commit-config.yaml, and pyproject.toml. The rest of the changes are the effect of applying the action to the current repo.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+repos:
+- repo: https://github.com/timothycrosley/isort
+  rev: 5.12.0
```
Comment (mc-nv) [inline .pre-commit-config.yaml:29]: No indentation for array

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  hooks:
+      - id: isort
```
Comment (mc-nv) [inline .pre-commit-config.yaml:31]: 4 spaces added as indentation

File: .pre-commit-config.yaml (hunk 1 above)
Comment (dyastremsky) [reply .pre-commit-config.yaml:29]: @mc-nv Should there be? I don't understand the goal here.

Comment (dyastremsky) [reply .pre-commit-config.yaml:29]: I think you mean you don't need the indent before "- id" below... got it, thanks. Will remove that.

Comment (dyastremsky) [reply .pre-commit-config.yaml:29]: Let me know if you're happy with the indentation changes. Can apply them to the other repos, if so.

File: .clang-format [hunk 3]
```diff
@@ -35,3 +36,4 @@ BinPackParameters: true
 ConstructorInitializerAllOnOneLineOrOnePerLine: false
 
 IndentCaseLabels: true
+
```
Comment (jbkyang-nvi) [inline .clang-format:39]: extra line here?

**Review by jbkyang-nvi — APPROVED** [review]

File: .clang-format (hunk 3 above)
Comment (dyastremsky) [reply .clang-format]: Thanks! Removed.

**Review by rmccorm4 — APPROVED** [review]

**Pull request merged**